aws-msk-iam SASL: credentials cache reconstructed on every handshake, causing STS throttling and request canceled failures under load

Summary
When using aws-msk-iam SASL with a roleArn (cross-account chained AssumeRole), every SASL handshake reconstructs the AWS SDK config, STS client, AssumeRoleProvider, and CredentialsCache from scratch. Because the cache is recreated each call, it is always cold — every SASL handshake triggers a fresh AssumeRoleWithWebIdentity (IRSA hop) plus a AssumeRole (chained hop) against STS.

At low resource counts this is invisible. Once concurrent SASL load is high enough to bump against STS rate limits, individual Retrieve() calls take long enough for franz-go's DialTimeout to trip, and the client logs:

[ERROR] unable to initialize sasl; broker: seed_0, err: request canceled, context canceled
…on roughly every reconcile attempt, with occasional success when timing aligns. From the outside it looks like a network or auth bug, but:

Broker kafka.server.socket_server_metrics.failed_authentication_total stays at zero (signed bytes never reach the broker — Retrieve doesn't return in time)
IAM "Last Accessed" on the IRSA role shows constant activity (STS is being hit constantly)
Existing non-IAM clients on the same listener authenticate normally
Direct franz-go usage from inside the same pod with the same SA succeeds end-to-end
Affected versions
provider-kafka v1.2.1 (confirmed by reading source; same pattern is likely in earlier versions that added aws-msk-iam support)
franz-go v1.21.0
aws-sdk-go-v2 config v1.32.16, credentials v1.19.15, service/sts v1.42.0
Root cause
[internal/clients/kafka/client.go](https://github.com/crossplane-contrib/provider-kafka/blob/v1.2.1/internal/clients/kafka/client.go) lines ~136-138:

```
case "aws-msk-iam":
    mechanism = kaws.ManagedStreamingIAM(func(ctx context.Context) (kaws.Auth, error) {
        return authenticateAwsIam(ctx, kc.SASL.RoleArn)
    })
```
authenticateAwsIam (called on every handshake):

```
func authenticateAwsIam(ctx context.Context, roleArn string) (a kaws.Auth, err error) {
    s, err := config.LoadDefaultConfig(ctx)            // fresh SDK config
    if err != nil { return kaws.Auth{}, err }
    if roleArn != "" {
        stsClient := sts.NewFromConfig(s)              // fresh STS client
        provider := stscreds.NewAssumeRoleProvider(...) // fresh provider
        s.Credentials = aws.NewCredentialsCache(provider) // FRESH empty cache
    }
    v, err := s.Credentials.Retrieve(ctx)              // always cold-miss
    ...
}
```

Every handshake builds a brand-new aws.NewCredentialsCache and immediately calls Retrieve on it. The cache is throwaway; credentials are never reused across handshakes within a single kgo client, let alone across reconciles.

Reproduction
Environment: EKS pod with IRSA → cross-account chained AssumeRole → MSK on the SASL_SSL IAM listener (port 9098). ~64 managed Topic resources spread across 4 ProviderConfig instances (4 different MSK clusters), single provider replica, default --max-reconcile-rate=10.

Symptoms:

Most Topic resources stuck at Synced=False with reason ReconcileError
Provider logs ~6 SASL errors/sec
Broker reports zero failed_authentication
IAM Last Accessed for the IRSA role: updated within last few minutes
Confirmed with a minimal direct franz-go program (same versions, same SA, same broker, same chained role) using a single long-lived aws.NewCredentialsCache: 64 concurrent goroutines × 60s of churn-and-retry = 4,747 successful Pings, 0 failures. The libraries are not the issue.

The bug also explains the partial per-PC success ratio: some PCs' reconciles happen to win the STS scheduling race more often than others, but none consistently succeed.

Proposed fix
Move the credentials-provider construction out of the per-handshake closure and into NewAdminClient, so the CredentialsCache lives for the lifetime of the kgo client (which is itself cached across reconciles by ClientCache keyed on the Secret hash).

```
@@ case "aws-msk-iam":
-    mechanism = kaws.ManagedStreamingIAM(func(ctx context.Context) (kaws.Auth, error) {
-        return authenticateAwsIam(ctx, kc.SASL.RoleArn)
-    })
+    awsCfg, err := config.LoadDefaultConfig(ctx)
+    if err != nil {
+        return nil, fmt.Errorf("cannot load AWS config: %w", err)
+    }
+    creds := awsCfg.Credentials
+    if kc.SASL.RoleArn != "" {
+        provider := stscreds.NewAssumeRoleProvider(sts.NewFromConfig(awsCfg), kc.SASL.RoleArn,
+            func(o *stscreds.AssumeRoleOptions) {
+                o.RoleSessionName = "crossplane-provider-kafka"
+            })
+        creds = aws.NewCredentialsCache(provider)
+    }
+    mechanism = kaws.ManagedStreamingIAM(func(ctx context.Context) (kaws.Auth, error) {
+        v, err := creds.Retrieve(ctx)
+        if err != nil {
+            return kaws.Auth{}, err
+        }
+        return kaws.Auth{
+            AccessKey:    v.AccessKeyID,
+            SecretKey:    v.SecretAccessKey,
+            SessionToken: v.SessionToken,
+            UserAgent:    "crossplane-provider-kafka",
+        }, nil
+    })
```

authenticateAwsIam becomes unused and can be removed.

With this change, STS calls drop from "2 per SASL handshake" (potentially hundreds per minute under reconcile load) to roughly "2 per credential lifetime" (one pair per hour per PC by default).

Workaround (no fork)
Cap reconcile concurrency via DeploymentRuntimeConfig so the unfixed cache reconstruction stays below STS throttling thresholds. For ~64 resources across 4 PCs:

```
spec:
  deploymentTemplate:
    spec:
      template:
        spec:
          containers:
          - name: package-runtime
            args:
            - --max-reconcile-rate=2
            - --poll-interval=2m
```
This restored all resources to Synced=True immediately in our environment, but it's a tourniquet — anyone scaling resource counts past the (cluster-specific) STS tipping point will hit the bug again.

--- a/README.md
+++ b/README.md
@@ -92,8 +92,19 @@ manage [Kafka](https://kafka.apache.org/) resources.
     the certificate/key from `clientCertificateSecretRef` will be used.
 
     **AWS MSK IAM**: When using `aws-msk-iam`, the provider uses the default AWS
-    credential chain (environment variables, IRSA, etc.). The IAM role needs at
-    minimum the following permissions to manage topics:
+    credential chain (environment variables, IRSA, etc.). For cross-account
+    access, set `roleArn` to chain an `AssumeRole` call on top of the default
+    credentials.
+
+    The optional `iamCredentialsExpiryWindow` field controls how early cached STS
+    credentials are refreshed **before** they expire (Go `time.Duration`
+    format, e.g. `"5m"`, `"30s"`). For example, with the default `"5m"` and
+    `AssumeRole`'s 15-minute credential lifetime, credentials are refreshed at
+    the 10-minute mark. Defaults to `"5m"`, maximum `"15m"`. Values that are
+    invalid or exceed the maximum are silently ignored and the default is used
+    instead.
+
+    The IAM role needs at minimum the following permissions to manage topics:
 
       ```json
       {

--- a/examples/namespaced/providerconfig/secret-aws-msk-iam.yaml.tmpl
+++ b/examples/namespaced/providerconfig/secret-aws-msk-iam.yaml.tmpl
@@ -13,7 +13,9 @@ stringData:
         "b-3.mycluster.abcdef.c1.kafka.us-east-1.amazonaws.com:9098"
       ],
       "sasl": {
-        "mechanism": "aws-msk-iam"
+        "mechanism": "aws-msk-iam",
+        "roleArn": "arn:aws:iam::123456789012:role/my-cross-account-role",
+        "iamCredentialsExpiryWindow": "5m"
       },
       "tls": {
         "insecureSkipVerify": false

--- a/internal/clients/kafka/client.go
+++ b/internal/clients/kafka/client.go
@@ -213,7 +213,7 @@ func configureFilePathCertificate(fr *ClientCertificatePath, tc *tls.Config) err
 // returned closure, so they persist across SASL handshakes for the lifetime
 // of the Kafka client. When a roleArn is configured, credentials are fetched
 // via chained AssumeRole with a configurable early-refresh window
-// (IAMCredentialsDuration) to avoid latency spikes at expiry.
+// (IAMCredentialsExpiryWindow) to avoid latency spikes at expiry.
 func newAwsMskIamMechanism(ctx context.Context, saslCfg *SASL) (sasl.Mechanism, error) {
 	awsCfg, err := config.LoadDefaultConfig(ctx)
 	if err != nil {
@@ -226,8 +226,8 @@ func newAwsMskIamMechanism(ctx context.Context, saslCfg *SASL) (sasl.Mechanism, 
 				o.RoleSessionName = "crossplane-provider-kafka"
 			})
 		expiryWindow := defaultIAMCredentialsExpiryWindow
-		if saslCfg.IAMCredentialsDuration != "" {
-			if d, err := time.ParseDuration(saslCfg.IAMCredentialsDuration); err == nil && d <= 15*time.Minute {
+		if saslCfg.IAMCredentialsExpiryWindow != "" {
+			if d, err := time.ParseDuration(saslCfg.IAMCredentialsExpiryWindow); err == nil && d <= 15*time.Minute {
 				expiryWindow = d
 			}
 		}

--- a/internal/clients/kafka/client.go
+++ b/internal/clients/kafka/client.go
@@ -58,7 +58,10 @@ func NewAdminClient(ctx context.Context, data []byte, kube client.Client) (*kadm
 	}
 
 	if kc.SASL != nil {
-		var mechanism sasl.Mechanism
+		var (
+			mechanism sasl.Mechanism
+			err       error
+		)
 		switch name := kc.SASL.Mechanism; strings.ToLower(name) {
 		case "plain":
 			mechanism = plain.Auth{
@@ -66,9 +69,10 @@ func NewAdminClient(ctx context.Context, data []byte, kube client.Client) (*kadm
 				Pass: kc.SASL.Password,
 			}.AsMechanism()
 		case "aws-msk-iam":
-			mechanism = kaws.ManagedStreamingIAM(func(ctx context.Context) (kaws.Auth, error) {
-				return authenticateAwsIam(ctx, kc.SASL.RoleArn)
-			})
+			mechanism, err = newAwsMskIamMechanism(ctx, kc.SASL)
+			if err != nil {
+				return nil, err
+			}
 		case "scram-sha-512":
 			mechanism = scram.Auth{
 				User: kc.SASL.Username,
@@ -116,36 +120,8 @@ func NewAdminClient(ctx context.Context, data []byte, kube client.Client) (*kadm
 	return kadm.NewClient(c), nil
 }
 
-func authenticateAwsIam(ctx context.Context, roleArn string) (a kaws.Auth, err error) {
-	s, err := config.LoadDefaultConfig(ctx)
-	if err != nil {
-		return kaws.Auth{}, err
-	}
-
-	if roleArn != "" {
-		stsClient := sts.NewFromConfig(s)
-		provider := stscreds.NewAssumeRoleProvider(stsClient, roleArn, func(o *stscreds.AssumeRoleOptions) {
-			o.RoleSessionName = "crossplane-provider-kafka"
-		})
-		s.Credentials = aws.NewCredentialsCache(provider)
-	}
-
-	v, err := s.Credentials.Retrieve(ctx)
-	if err != nil {
-		return kaws.Auth{}, err
-	}
-
-	a = kaws.Auth{
-		AccessKey:    v.AccessKeyID,
-		SecretKey:    v.SecretAccessKey,
-		SessionToken: v.SessionToken,
-		UserAgent:    "crossplane-provider-kafka",
-	}
-
-	return a, nil
-}
-
-// Add options to TLS config for client certificate (if configured)
+// configureClientCertificate sets up client certificate authentication in the TLS config,
+// supporting both Kubernetes Secret references and on-disk file paths.
 func configureClientCertificate(ctx context.Context, kc Config, kube client.Client, tc *tls.Config) error {
 	// Validate that both clientCertificateSecretRef and clientCertificatePath are not both set.
 	// In Go TLS, GetClientCertificate (used for file-based certs) takes precedence over Certificates (used for secret-based certs),
@@ -167,6 +143,8 @@ func configureClientCertificate(ctx context.Context, kc Config, kube client.Clie
 	return configureCACertificateFile(kc.TLS.CACertificateFile, tc)
 }
 
+// configureSecretRefCertificate loads a client certificate key pair from a Kubernetes Secret
+// and appends it to the TLS config.
 func configureSecretRefCertificate(ctx context.Context, sr *ClientCertificateSecretRef, kube client.Client, tc *tls.Config) error {
 	if sr == nil {
 		return nil
@@ -205,6 +183,8 @@ func configureSecretRefCertificate(ctx context.Context, sr *ClientCertificateSec
 	return nil
 }
 
+// configureFilePathCertificate sets up TLS client certificate from on-disk files with
+// automatic rotation support via the GetClientCertificate callback.
 func configureFilePathCertificate(fr *ClientCertificatePath, tc *tls.Config) error {
 	if fr == nil {
 		return nil
@@ -228,6 +208,49 @@ func configureFilePathCertificate(fr *ClientCertificatePath, tc *tls.Config) err
 	return nil
 }
 
+// newAwsMskIamMechanism builds a SASL mechanism for AWS MSK IAM authentication.
+// AWS config and credentials cache are constructed once and captured by the
+// returned closure, so they persist across SASL handshakes for the lifetime
+// of the Kafka client. When a roleArn is configured, credentials are fetched
+// via chained AssumeRole with a configurable early-refresh window
+// (IAMCredentialsDuration) to avoid latency spikes at expiry.
+func newAwsMskIamMechanism(ctx context.Context, saslCfg *SASL) (sasl.Mechanism, error) {
+	awsCfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load AWS config: %w", err)
+	}
+	creds := awsCfg.Credentials
+	if saslCfg.RoleArn != "" {
+		provider := stscreds.NewAssumeRoleProvider(sts.NewFromConfig(awsCfg), saslCfg.RoleArn,
+			func(o *stscreds.AssumeRoleOptions) {
+				o.RoleSessionName = "crossplane-provider-kafka"
+			})
+		expiryWindow := defaultIAMCredentialsExpiryWindow
+		if saslCfg.IAMCredentialsDuration != "" {
+			if d, err := time.ParseDuration(saslCfg.IAMCredentialsDuration); err == nil && d <= 15*time.Minute {
+				expiryWindow = d
+			}
+		}
+		creds = aws.NewCredentialsCache(provider,
+			func(o *aws.CredentialsCacheOptions) {
+				o.ExpiryWindow = expiryWindow
+			})
+	}
+	return kaws.ManagedStreamingIAM(func(ctx context.Context) (kaws.Auth, error) {
+		v, err := creds.Retrieve(ctx)
+		if err != nil {
+			return kaws.Auth{}, err
+		}
+		return kaws.Auth{
+			AccessKey:    v.AccessKeyID,
+			SecretKey:    v.SecretAccessKey,
+			SessionToken: v.SessionToken,
+			UserAgent:    "crossplane-provider-kafka",
+		}, nil
+	}), nil
+}
+
+// validateClientCertificatePath checks that cert and key files exist on disk and form a valid pair.
 func validateClientCertificatePath(fr *ClientCertificatePath) error {
 	// Validate that files exist and can be read initially
 	if _, err := os.Stat(fr.CertFile); err != nil {
@@ -253,6 +276,8 @@ func validateClientCertificatePath(fr *ClientCertificatePath) error {
 	return nil
 }
 
+// loadClientCertificate reads a certificate key pair from disk, called on each TLS handshake
+// to support certificate rotation without restart.
 func loadClientCertificate(fr *ClientCertificatePath) (*tls.Certificate, error) {
 	certPEM, err := os.ReadFile(fr.CertFile)
 	if err != nil {
@@ -271,6 +296,8 @@ func loadClientCertificate(fr *ClientCertificatePath) (*tls.Certificate, error) 
 	return &kp, nil
 }
 
+// configureCACertificateSecretRef loads a CA certificate from a Kubernetes Secret and appends
+// it to the TLS root CA pool.
 func configureCACertificateSecretRef(ctx context.Context, sr *CACertificateSecretRef, kube client.Client, tc *tls.Config) error {
 	if sr == nil {
 		return nil
@@ -292,6 +319,8 @@ func configureCACertificateSecretRef(ctx context.Context, sr *CACertificateSecre
 	return appendCACert(caPEM, tc)
 }
 
+// configureCACertificateFile loads a CA certificate from a file path and appends it to the
+// TLS root CA pool.
 func configureCACertificateFile(caFile string, tc *tls.Config) error {
 	if caFile == "" {
 		return nil
@@ -305,6 +334,8 @@ func configureCACertificateFile(caFile string, tc *tls.Config) error {
 	return appendCACert(caPEM, tc)
 }
 
+// appendCACert appends a PEM-encoded CA certificate to the TLS root CA pool,
+// initializing from system roots if the pool does not yet exist.
 func appendCACert(caPEM []byte, tc *tls.Config) error {
 	var pool *x509.CertPool
 

--- a/internal/clients/kafka/config.go
+++ b/internal/clients/kafka/config.go
@@ -32,11 +32,12 @@ type Config struct {
 type SASL struct {
 	Mechanism string `json:"mechanism"`
 	RoleArn   string `json:"roleArn"`
-	// IAMCredentialsDuration is the early-refresh window for cached AWS credentials.
-	// Uses Go time.Duration format (e.g. "5m", "30s", "2m30s"). Defaults to "5m".
-	IAMCredentialsDuration string `json:"iamCredentialsDuration,omitempty"`
-	Username               string `json:"username"`
-	Password               string `json:"password"` //nolint:gosec
+	// IAMCredentialsExpiryWindow controls how early cached STS credentials are refreshed
+	// before expiry. Uses Go time.Duration format (e.g. "5m", "30s"). Defaults to "5m",
+	// maximum "15m".
+	IAMCredentialsExpiryWindow string `json:"iamCredentialsExpiryWindow,omitempty"`
+	Username                   string `json:"username"`
+	Password                   string `json:"password"` //nolint:gosec
 }
 
 // TLS is an option for enabling encryption in transit.

--- a/internal/clients/kafka/config.go
+++ b/internal/clients/kafka/config.go
@@ -32,8 +32,11 @@ type Config struct {
 type SASL struct {
 	Mechanism string `json:"mechanism"`
 	RoleArn   string `json:"roleArn"`
-	Username  string `json:"username"`
-	Password  string `json:"password"` //nolint:gosec
+	// IAMCredentialsDuration is the early-refresh window for cached AWS credentials.
+	// Uses Go time.Duration format (e.g. "5m", "30s", "2m30s"). Defaults to "5m".
+	IAMCredentialsDuration string `json:"iamCredentialsDuration,omitempty"`
+	Username               string `json:"username"`
+	Password               string `json:"password"` //nolint:gosec
 }
 
 // TLS is an option for enabling encryption in transit.

--- a/internal/clients/kafka/constants.go
+++ b/internal/clients/kafka/constants.go
@@ -1,8 +1,13 @@
 package kafka
 
-import "crypto/tls"
+import (
+	"crypto/tls"
+	"time"
+)
 
 const (
+	defaultIAMCredentialsExpiryWindow = 5 * time.Minute
+
 	// ACL resource types
 	ACLResourceTypeTopic           = "Topic"
 	ACLResourceTypeGroup           = "Group"

